aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -45,18 +45,19 @@ static inline Tensor& unary_op_impl_out(Tensor& result, const Tensor& self, Stub
 }
 
 template <typename Stub>
-static inline Tensor& unary_op_impl_float_out(Tensor& result, const Tensor& self, Stub& stub) {
-  auto iter = TensorIterator::unary_float_op(result, self);
+static inline Tensor& unary_op_impl_float_out(Tensor& result, const Tensor& self, Stub& stub, c10::optional<ScalarType> dtype = c10::nullopt) {
+  auto iter = TensorIterator::unary_float_op(result, self, dtype);
   stub(iter.device_type(), iter);
   iter.cast_outputs();
   return result;
 }
 
 template <typename Stub>
-Tensor unary_op_impl_float(const Tensor& self, Stub& stub) {
+Tensor unary_op_impl_float(const Tensor& self, Stub& stub, c10::optional<ScalarType> dtype = c10::nullopt) {
   Tensor result;
-  auto iter = TensorIterator::unary_float_op(result, self);
+  auto iter = TensorIterator::unary_float_op(result, self, dtype);
   stub(iter.device_type(), iter);
+  iter.cast_outputs();
```

Comment:
Was this a bug? If so, are there tests we can add that would have caught it?

Reply:
Follow-up question: should this just call `unary_op_impl_float_out()`?

Reply:
> Was this a bug? If so, are there tests we can add that would have caught it?

I don't think so. It works fine without that. Added it while experimenting.

> Follow-up question: should this just call unary_op_impl_float_out()?

I had previously tried similar while adding the utility `unary_op_impl_float` and tried even now,
```C++
template <typename Stub>
Tensor unary_op_impl_float(const Tensor& self, Stub& stub, c10::optional<ScalarType> dtype = c10::nullopt) {
  Tensor result;
  return unary_op_impl_float_out(result, self, stub, dtype);
}
```

But that results in returning `None` instead of Tensor
```python
>>> import torch
>>> t = torch.randn(3)
>>> t.sin()
>>> r = t.sin()
>>> r is None
True
```

Do note that similar pattern is used in BinaryOps
https://github.com/pytorch/pytorch/blob/f7b339d11cf03edac957ce4cbaa2a9edfec3caf1/aten/src/ATen/native/BinaryOps.cpp#L404-L436

Never dug up to see the exact details. It does look a bit fishy. Worth investigating?